File: submit-web/src/models/Package.ts
import { SubmissionItem } from "./SubmissionItem";
import { UpdateRequest } from "./UpdateRequest";

export type PackageType = {
  id: number;
  name: string;
};

// These statuses are just for UI purposes, the actual canonical business statuses are PackageStatus
export type NonCanonicalPackageStatus =
  | "REVISION_REQUIRED"
  | "UPDATED"
  | "NO_REVISION_REQUIRED"
  | "RESUBMITTED"
  | "UPDATE_REQUESTED";

export const NON_CANONICAL_PACKAGE_STATUS = Object.freeze<
  Record<NonCanonicalPackageStatus, NonCanonicalPackageStatus>
>({
  UPDATE_REQUESTED: "UPDATE_REQUESTED",
  REVISION_REQUIRED: "REVISION_REQUIRED",
  UPDATED: "UPDATED",
  NO_REVISION_REQUIRED: "NO_REVISION_REQUIRED",
  RESUBMITTED: "RESUBMITTED",
});

export type PackageStatus =
  | "IN_REVIEW"
  | "UNDER_REVIEW"
  | "APPROVED"
  | "ACCEPTED"
  | "SATISFIED"
  | "REVIEW_REJECTED"
  | "REVIEWED"
  | "COMPLETED"
  | "SUBMITTED"
  | "PARTIALLY_COMPLETED"
  | "NEW_SUBMISSION"
  | "UNDER_CONSULTATION_CHECK"
  | "PASSED_CONSULTATION_CHECK"
  | "AWAITING_MANAGER_APPROVAL"
  | "FAILED_CONSULTATION_CHECK"
  | "CREATED";

export const PACKAGE_STATUS: Record<
  PackageStatus,
  { value: PackageStatus; label: string }
> = {
  IN_REVIEW: {
    value: "IN_REVIEW",
    label: "In Review",
  },
  UNDER_REVIEW: {
    value: "UNDER_REVIEW",
    label: "Under Review",
  },
  APPROVED: {
    value: "APPROVED",
    label: "Approved",
  },
  REVIEW_REJECTED: {
    value: "REVIEW_REJECTED",
    label: "Rejected",
  },
  COMPLETED: {
    value: "COMPLETED",
    label: "Completed",
  },
  SUBMITTED: {
    value: "SUBMITTED",
    label: "Submitted",
  },
  PARTIALLY_COMPLETED: {
    value: "PARTIALLY_COMPLETED",
    label: "Partially Completed",
  },
  NEW_SUBMISSION: {
    value: "NEW_SUBMISSION",
    label: "New Submission",
  },
  UNDER_CONSULTATION_CHECK: {
    value: "UNDER_CONSULTATION_CHECK",
    label: "Under Consultation Check",
  },
  PASSED_CONSULTATION_CHECK: {
    value: "PASSED_CONSULTATION_CHECK",
    label: "Passed Consultation Check",
  },
  CREATED: {
    value: "CREATED",
    label: "Created",
  },
  ACCEPTED: {
    value: "ACCEPTED",
    label: "Accepted",
  },
  SATISFIED: {
    value: "SATISFIED",
    label: "Satisfied",
  },
  AWAITING_MANAGER_APPROVAL: {
    value: "AWAITING_MANAGER_APPROVAL",
    label: "Awaiting Manager Approval",
  },
  FAILED_CONSULTATION_CHECK: {
    value: "FAILED_CONSULTATION_CHECK",
    label: "Failed Consultation Check",
  },
  REVIEWED: {
    value: "REVIEWED",
    label: "Reviewed",
  },
};

export type SubmissionPackageMeta = Record<string, any>;

export type PackageVersion = {
  id: number;
  package_id: number;
  version: number;
  original_package_id: number;
  is_approved: boolean;
};

export type SubmissionPackage = {
  id: number;
  name: string;
  status: PackageStatus[];
  submitted_on?: string;
  completed_on?: string;
  submitted_by?: string;
  type_id: number;
  type: PackageType;
  items: Array<SubmissionItem>;
  account_project_id: number;
  meta?: SubmissionPackageMeta;
  days_since_submission?: number;
  review_status?: NonCanonicalPackageStatus;
  update_requests: Array<UpdateRequest>;
  version: PackageVersion;
};

File: submit-api/src/submit_api/models/package.py
"""Submission package model class.

Manages the package
"""
from __future__ import annotations

import enum

from sqlalchemy import Column, Enum, ForeignKey
from sqlalchemy.orm import joinedload

from .base_model import BaseModel
from .db import db


class PackageStatus(enum.Enum):
    """Enum for package statuses."""

    IN_REVIEW = 'IN_REVIEW'
    APPROVED = 'APPROVED'
    ACCEPTED = 'ACCEPTED'
    SATISFIED = 'SATISFIED'
    REJECTED = 'REJECTED'
    REVIEWED = 'REVIEWED'
    SUBMITTED = 'SUBMITTED'
    PARTIALLY_COMPLETED = 'PARTIALLY_COMPLETED'
    COMPLETED = 'COMPLETED'
    NEW_SUBMISSION = 'NEW_SUBMISSION'
    PASSED_CONSULTATION_CHECK = 'PASSED_CONSULTATION_CHECK'
    FAILED_CONSULTATION_CHECK = 'FAILED_CONSULTATION_CHECK'
    UNDER_REVIEW = 'UNDER_REVIEW'
    UNDER_CONSULTATION_CHECK = 'UNDER_CONSULTATION_CHECK'
    REVIEW_REJECTED = 'REVIEW_REJECTED'
    CREATED = 'CREATED'
    AWAITING_MANAGER_APPROVAL = 'AWAITING_MANAGER_APPROVAL'
    CC_AWAITING_MANAGER_APPROVAL = 'CC_AWAITING_MANAGER_APPROVAL'
    MP_AWAITING_MANAGER_APPROVAL = 'MP_AWAITING_MANAGER_APPROVAL'
    REVISION_REQUIRED = 'REVISION_REQUIRED'
    NO_REVISION_REQUIRED = 'NO_REVISION_REQUIRED'
    RESUBMITTED = 'RESUBMITTED'


class Package(BaseModel):
    """Definition of the package entity."""

    __tablename__ = 'packages'

    id = Column(db.Integer, primary_key=True, autoincrement=True)
    account_project_id = Column(db.Integer, ForeignKey(
        'account_projects.id'), nullable=False)
    name = Column(db.String(255), nullable=False)
    type_id = Column(db.Integer, ForeignKey(
        'package_types.id'), nullable=False)
    type = db.relationship('PackageType', foreign_keys=[
                           type_id], lazy='joined')
    submitted_on = Column(db.DateTime, nullable=True)
    submitted_by = Column(db.String, ForeignKey(
        'users.auth_guid'), nullable=True)
    submitted_by_user = db.relationship(
        'User', foreign_keys=[submitted_by], lazy='joined')
    completed_on = Column(db.DateTime, nullable=True)
    meta = db.relationship(
        'PackageMetadata', backref='package', lazy='joined', uselist=False)
    items = db.relationship('Item', backref='package',
                            lazy='select', order_by='Item.sort_order')
    status = Column(db.ARRAY(Enum(PackageStatus)), nullable=False,
                    default=[PackageStatus.NEW_SUBMISSION.value])
    active = Column(db.Boolean, nullable=False, default=True)
    version_id = Column(db.Integer, ForeignKey(
        'package_versions.id'), nullable=True)
    version = db.relationship('PackageVersion', foreign_keys=[
                              version_id], lazy='joined')

    _update_requests = db.relationship(
        'UpdateRequest',
        backref='package',
        lazy='joined')

    @property
    def update_requests(self):
        """Get the active update requests for the package."""
        return [ur for ur in self._update_requests if ur.active]

    @property
    def all_update_requests(self):
        """Get all update requests for the package."""
        return self._update_requests

    @classmethod
    def get_package_by_id_with_items(cls, package_id: int):
        """Return model by package id."""
        return cls.query.filter_by(id=package_id).options(joinedload(Package.items)).first()

    @classmethod
    def get_all_package_by_ids(cls, package_ids: list[int]):
        """Return model by package ids."""
        return cls.query.filter(Package.id.in_(package_ids)).all()
